### ISSUE TITLE
Ellipses for parameters in C#

### ISSUE THREAD

**Issue by Sjord** [semgrep/semgrep#3289]
It would be nice to match any C# method like this:

```
public $A SomeMethod(...) { ... }
```

For this to work, ellipses should match zero or more parameters.

I tried adding the following to [fyi/semgrep-grammars/src/semgrep-c-sharp/grammar.js](https://github.com/returntocorp/semgrep-c-sharp/blob/main/fyi/semgrep-grammars/src/semgrep-c-sharp/grammar.js).

```
    parameter: ($, previous) => {
      return choice(
        ...previous.members,
        $.ellipsis
      );
    },
```

However, this is overwritten on build. Is this the correct method? How do I modify this file in a manner that it is included in the build?

**Comment by aryx** [comment]
This is the one you want to modify: https://github.com/returntocorp/ocaml-tree-sitter/blob/master/lang/semgrep-grammars/src/semgrep-c-sharp/grammar.js

It's explained a bit here: https://github.com/returntocorp/ocaml-tree-sitter/blob/master/doc/updating-a-grammar.md
"syntax extensions to support semgrep patterns, such as ellipses (...) and metavariables ($FOO). This is lang/semgrep-grammars/src/semgrep-X. It can be tested from that folder with make && make test. 
..."

**Comment by Sjord** [comment]
I [attempted to implement this](https://github.com/Sjord/semgrep-c-sharp/commit/df62e0cca53b3734c46333ea019d7c4306565025). I haven't tested it in semgrep yet. A problem I ran into is that the definition for `interpolated_string_text` changed in CST.ml.

It was this:
```
type interpolated_string_text = [
    `LCURLLCURL of Token.t (* "{{" *)
  | `Imm_tok_pat_2755817 of imm_tok_pat_2755817 (*tok*)
  | `Esc_seq of escape_sequence (*tok*)
]
```

And now it's this:
```
type interpolated_string_text = [
    `LCURLLCURL of Token.t (* "{{" *)
  | `Blank of unit (* blank *)
  | `Esc_seq of escape_sequence (*tok*)
]
```

I am not sure what is going on.

**Comment by Sjord** [comment]
@mjambon Is this the result of your recent changes to ocaml-tree-sitter?

The C# grammar has this:

```
    interpolated_string_text: $ => choice(
      '{{',
      token.immediate(prec(1, /[^{"\\\n]+/)),
      $.escape_sequence
    ),
```

and that is translated in CST.ml to this:
```
type interpolated_string_text = [
    `LCURLLCURL of Token.t (* "{{" *)
  | `Blank of unit (* blank *)
  | `Esc_seq of escape_sequence (*tok*)
]
```

So the second entry, `token.immediate(...)`, is always mapped to `Blank`, which seems incorrect. Is the C# grammar incorrect? What should it be? Or is this a bug in ocaml-tree-sitter?

**Comment by mjambon** [comment]
@Sjord yes, this is the result of the changes I made last week in ocaml-tree-sitter as the refinement of a previous workaround for missing tokens in the tree-sitter CST. I may have made a wrong assumption for this particular case.

In the tree-sitter-hack grammar, [we have a case of a `token(prec.dynamic(...))`](https://github.com/antoniodejesusochoasolano/tree-sitter-hack/blob/7029040f29a1a451707edbbfdfa985d809520e79/grammar.js#L293), and the old behavior was to create a new rule, like `imm_tok_pat_2755817`, as a workaround to obtain an actual token in the CST. The problem is that creating a new rule was causing a parsing conflict to be reported, with no obvious automatic fix. I can't tell if that's normal or if it's a bug in tree-sitter. So, I decided to make an exception and not extract `token()` into a new rule if they contain precedence values. And I did that for all the "prec" constructs: `prec`, `prec.left`, `prec.right`, and `prec.dynamic`. Maybe I should have done this only for `prec.dynamic`, since it looks like it was not a problem for `token.immediate(prec(1, /[^{"\\\n]+/))`.

The whole issue with missing tokens is [described there](https://github.com/tree-sitter/tree-sitter/issues/1156). Note that some constructs like `token(repeat('x')` cause memory explosion and eventual crashing at parsing time. My advice to grammar writers would be to stay away from `token()` as much as possible for the time being. I understand this may not be very realistic.

Note that I don't understand what the `prec` or `prec.dynamic` within a `token()` are meant to achieve and what automatic transformations of the grammar would be safe. I would need to know when extracting `token(prec(...))` into its own rule (resp. `token(prec.dynamic(...))`) is ok and when it's not ok. In the meantime, I'll err on the conservative side.

My best suggestion for now is to rewrite the grammar manually in tree-sitter-c-sharp and create a named rule for the `token.immediate()`:
```js
    interpolated_string_text: $ => choice(
      '{{',
      token.immediate(prec(1, /[^{"\\\n]+/)),
      $.escape_sequence
    ),

```
would become something like
```js
    interpolated_string_text: $ => choice(
      '{{',
      $.string_fragment,
      $.escape_sequence
    ),
    string_fragment: $ => token.immediate(prec(1, /[^{"\\\n]+/)),
```
As small benefit, you'd also get a nice name for the token rather than `imm_tok_pat_2755817`.

The grammar [can also be patched in ocaml-tree-sitter/lang](https://github.com/returntocorp/ocaml-tree-sitter/blob/master/lang/semgrep-grammars/src/semgrep-c-sharp/grammar.js) since we support grammar extensions. This is meant to support semgrep constructs like `...` and such which shouldn't be part of the official language grammar. I would only recommend the extension mechanism as a last resort, though.

**Comment by Sjord** [comment]
[Extract _string_literal_fragment and _interpolated_string_text_fragment by Sjord · Pull Request #196 · tree-sitter/tree-sitter-c-sharp](https://github.com/tree-sitter/tree-sitter-c-sharp/pull/196)

> Note that I don't understand what the prec or prec.dynamic within a token() are meant to achieve

Apparently it is to [move precedence handling to the lexer](https://github.com/tree-sitter/tree-sitter/issues/1087#issuecomment-833198651).

**Comment by inkz** [comment]
@aryx @mjambon 
will this also tackle situation like this:

```
public $A SomeMethod(..., $X, ...) { ... }
```

or I should open another ticket?

**Comment by mjambon** [comment]
@inkz 
* If you're referring to the return type captured by `$A`, it looks like it should be already supported since `identifier` is extended to support semgrep metavariables, and an `identifier` is valid in this position.
* Same reasoning for `$X`. The type appears to be optional, so I'm guessing a plain `$X` should be valid for a `parameter`.

Here's where to look:
* The grammar extension for semgrep is here: https://github.com/returntocorp/ocaml-tree-sitter/blob/master/lang/semgrep-grammars/src/semgrep-c-sharp/grammar.js
* The original tree-sitter-c-sharp grammar is https://github.com/tree-sitter/tree-sitter-c-sharp/blob/master/grammar.js

FYI last week I learned that on github we can just click on grammar rules in grammar.js to see and/or visit their definition, it's handy!

**Comment by mjambon** [comment]
@Sjord I'm noticing [this](https://github.com/returntocorp/ocaml-tree-sitter/blob/5c17d313d9ac8c55ad59410e2439a6b6cbc29c34/lang/semgrep-grammars/src/semgrep-c-sharp/grammar.js#L43) in the grammar extension:
![image](https://user-images.githubusercontent.com/343265/122820649-b2b10000-d290-11eb-912f-46d0140ccd27.png)

You may prefer to create a rule for the pattern (regexp) since otherwise it will be automatically created with an obscure name e.g. do this:
```js
semgrep_metavariable: $ => /\$[A-Z_][A-Z_0-9]*/
```

**Comment by inkz** [comment]
@mjambon thank you:+1:

**Comment by Sjord** [comment]
My [ocaml-tree-sitter PR was merged](https://github.com/returntocorp/ocaml-tree-sitter/pull/218). I need someone at R2C to release a new semgrep-c-sharp. Then I can refer to that in [this PR](https://github.com/returntocorp/semgrep/pull/3412) and then ellipsis can be used as parameters.

**Comment by mjambon** [comment]
@Sjord it's ready: https://github.com/returntocorp/semgrep-c-sharp/tree/dd52ab7e924556b3c500ad1c13232748c0942d00

**Comment by Sjord** [comment]
@mjambon [Something went wrong](https://github.com/returntocorp/semgrep-c-sharp/blob/dd52ab7e924556b3c500ad1c13232748c0942d00/lib/CST.ml#L174):

```
type interpolated_string_text = [
    `LCURLLCURL of Token.t (* "{{" *)
  | `Blank of unit (* blank *)
  | `Esc_seq of escape_sequence (*tok*)
]
```

The text fragment is `Blank` again. Did you use the latest upstream tree-sitter-c-sharp grammar?

**Comment by mjambon** [comment]
Oops. It should be good now.
Before: https://github.com/returntocorp/semgrep-c-sharp/blob/dd52ab7e924556b3c500ad1c13232748c0942d00/fyi/semgrep-grammars/src/tree-sitter-c-sharp/grammar.js#L1161-L1165
After: https://github.com/returntocorp/semgrep-c-sharp/blob/06f15cd5de732f560f3a14e6f43c1baca73e4767/fyi/semgrep-grammars/src/tree-sitter-c-sharp/grammar.js#L1165-L1169

**Comment by mjambon** [comment]
no more blank in `CST.ml`:
https://github.com/returntocorp/semgrep-c-sharp/blob/06f15cd5de732f560f3a14e6f43c1baca73e4767/lib/CST.ml#L212-L216
```ocaml
type interpolated_string_text = [
    `LCURLLCURL of Token.t (* "{{" *)
  | `Inte_str_text_frag of interpolated_string_text_fragment (*tok*)
  | `Esc_seq of escape_sequence (*tok*)
]
```

**Comment by Sjord** [comment]
Thanks, the [pull request](https://github.com/returntocorp/semgrep/pull/3412) works and is ready for review.

**Comment by Sjord** [comment]
Pull request is merged, so this issue is done. Available either in the current `develop` branch or with the next release.

@inkz, please test, and feel free to file a new issue if your use case is not solved by this change.

**Comment by inkz** [comment]
@Sjord ok, thank you :+1:
